src/main/java/uk/gov/hmcts/darts/cache/token/service/AbstractTokenCache.java

```diff
@@ -299,6 +182,63 @@ private Optional<Token> getTokenValueWithResetExpiration(CacheValue holder) {
         return Optional.empty();
     }
 
+
+    @SuppressWarnings("java:S6809")
+    private Optional<Token> storeIntoRedis(CacheValue value, Boolean reuseTokenIfPossible, boolean validateToken) throws CacheException {
+
+        log.debug("Storing the supplied value");
+
+        Optional<Token> tokenToUse = Optional.empty();
+
+        // if we have found a token, but it is invalid then invalidate it and force a new token to be created
+        if (reuseTokenBasedOnCredentials(reuseTokenIfPossible)) {
+            log.debug("Looking up token in cache");
+
+            tokenToUse = read(value);
+
+            if (tokenToUse.isPresent() && validateToken) {
+                log.debug("Found the token in the cache");
+                if (!tokenToUse.get().validate(Token.TokenExpiryEnum.APPLY_EARLY_TOKEN_EXPIRY)) {
+                    tokenToUse = Optional.empty();
+                }
+            } else {
+                log.debug("Token not found in cache");
+            }
+
+            if (tokenToUse.isEmpty()) {
+                // ensure that one token is acquired when we know one is needed
+                CacheLockableUnitOfWork distributedLockWork = new CacheLockableUnitOfWork(lockRegistry);
+                tokenToUse = distributedLockWork.execute(() -> {
+
+                    // read the token again one at a time to skip token creation
+                    Optional<Token> foundToken = read(value);
+                    if (foundToken.isPresent() && !foundToken.get().validate(Token.TokenExpiryEnum.APPLY_EARLY_TOKEN_EXPIRY)) {
```

Comment:
conscious that we potentially only just validated this token on line 197->201.
Would it be easy to check if its the same token instead of revalidating?
Other than that, it does something very similar to lines 197->201, albeit not taking validateToke into account, so see previous comment about making into its own method.